gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -447,16 +537,30 @@ private ManagedChannel createSingleChannel() throws IOException {
       builder.keepAliveTime(DIRECT_PATH_KEEP_ALIVE_TIME_SECONDS, TimeUnit.SECONDS);
       builder.keepAliveTimeout(DIRECT_PATH_KEEP_ALIVE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
     } else {
+      // Try and create credentials via DCA. See https://google.aip.dev/auth/4114.
       ChannelCredentials channelCredentials;
       try {
         channelCredentials = createMtlsChannelCredentials();
       } catch (GeneralSecurityException e) {
         throw new IOException(e);
       }
       if (channelCredentials != null) {
+        // Create the channel using channel credentials created via DCA.
         builder = Grpc.newChannelBuilder(endpoint, channelCredentials);
       } else {
-        builder = ManagedChannelBuilder.forAddress(serviceAddress, port);
+        // Could not create channel credentials via DCA. In accordance with
+        // https://google.aip.dev/auth/4115, if credentials not available through
+        // DCA, try mTLS with credentials held by the S2A (Secure Session Agent).
+        if ((endpointContext != null) && endpointContext.useS2A()) {
```

Comment:
Ah, ok I think I understand the `needsEndpointContext()` logic here. By default, an empty instance for EndpointContext is created in StubSettings: https://github.com/googleapis/sdk-platform-java/blob/25023afad8f483a3eae8846c4bc1fbe2c7a260c5/gax-java/gax/src/main/java/com/google/api/gax/rpc/StubSettings.java#L370

I don't think we need the endpointContext null check here.

Reply:
I think I added this check because some tests were not setting it, I removed it and mocked it in tests in https://github.com/googleapis/sdk-platform-java/pull/3326/commits/089773020f2324c8bb75dbc98954f3379dd28b56.